container/state.go

```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
 	return true
 }
 
-func wait(waitChan <-chan struct{}, timeout time.Duration) error {
-	if timeout < 0 {
-		<-waitChan
-		return nil
-	}
-	select {
-	case <-time.After(timeout):
-		return fmt.Errorf("Timed out: %v", timeout)
-	case <-waitChan:
-		return nil
-	}
-}
+// WaitCondition is an enum type for different states to wait for.
+type WaitCondition int
+
+// Possible WaitCondition Values.
+//
+// WaitConditionNotRunning (default) is used to wait for any of the non-running
+// states: "created", "exited", "dead", "removing", or "removed".
+//
+// WaitConditionNextExit is used to wait for the next time the state changes
+// to a non-running state. If the state is currently "created" or "exited",
+// this would cause Wait() to block until either the container runs and exits
+// or is removed.
+//
+// WaitConditionRemoved is used to wait for the container to be removed.
+const (
+	WaitConditionNotRunning WaitCondition = iota
+	WaitConditionNextExit
+	WaitConditionRemoved
+)
+
+// Wait waits until the continer is in a certain state indicated by the given
+// condition. A context can be used for cancelling the request or controlling
+// timeouts. Wait must be called without holding the state lock. Returns a
+// channel which can be used to receive the result. If the container exited on
+// its own, the result's Err() method wil be nil and its ExitCode() method will
+// return the conatiners exit code, otherwise, the results Err() method will
+// return an error indicating why the wait operation failed.
+func (s *State) Wait(ctx context.Context, condition WaitCondition) <-chan *StateStatus {
```

Comment:
> So if it didn't return a channel then it would have to block? Wouldn't you still want to pass in a context to cancel that?

Canceling a context if a caller doesn't need a result anymore is what the `context.WithCancel` is meant for. Requiring in the implementation that a context passed into the function must be canceled in order to not leak resources is a different constraint.

This `Wait` has no way to return early and provide an early error. If it is more valuable than confusing that you can have an error with `200` status code that doesn't need to leak into lower-level functions. If I look at the client side code(and builder)
https://github.com/moby/moby/pull/32237/files#diff-316ca96dd5834654f8dc1537e5377bc3R28 then I don't see this property of HTTP api used. I don't any callers checking the status of the returned channel and error channel differently. So do we actually handle that race?